pulsar-function-go/pf/instance.go

```diff
@@ -0,0 +1,256 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package pf
+
+import (
+	"context"
+	"time"
+
+	"github.com/apache/pulsar/pulsar-client-go/pulsar"
+	"github.com/apache/pulsar/pulsar-function-go/log"
+	"github.com/apache/pulsar/pulsar-function-go/pb"
+)
+
+type GoInstance struct {
+	function  Function
+	context   *FunctionContext
+	producer  pulsar.Producer
+	consumers map[string]pulsar.Consumer
+	client    pulsar.Client
+}
+
+// NewGoInstance init GoInstance and init function context
+func NewGoInstance() *GoInstance {
+	goInstance := &GoInstance{
+		context:   NewFuncContext(),
+		consumers: make(map[string]pulsar.Consumer),
+	}
+	return goInstance
+}
+
+func (gi *GoInstance) StartFunction(function Function) {
+	gi.function = function
+	gi.setupClient()
+	err := gi.setupProducer()
+	if err != nil {
+		log.Errorf("setup producer failed, error is:%s", err.Error())
+	}
+	channel := gi.setupConsumer()
+
+CLOSE:
+	for {
+		select {
+		case cm := <-channel:
+			msgInput := cm.Message
+			output, err := gi.handlerMsg(msgInput)
+			if err != nil {
+				log.Errorf("handler message error:%v", err)
+			}
+			err = gi.processResult(msgInput, output)
+			if err != nil {
+				log.Errorf("process result error:%v", err)
+			}
+		case <-time.After(time.Millisecond * gi.context.InstanceConf.KillAfterIdleMs):
```

Comment:
you need check if `KillAfterIdleMs` is larger than `0`. if it is a negative value, it should be treated as infinite waiting until the instance was terminated or killed.